src/cryptonote_protocol/quorumnet.cpp

```diff
@@ -1364,6 +1392,233 @@ void handle_blink_success(Message& m) {
+
+    msg.type             = pulse::message_type::handshake_bitset;
+    msg.validator_bitset = validator_bitset;
+  }
+  else
+  {
+    auto buf = tools::memcpy_le(top_hash.data, msg.quorum_position);
+    crypto::hash hash;
+    crypto::cn_fast_hash(buf.data(), buf.size(), hash);
+    crypto::generate_signature(hash, core.get_service_keys().pub, core.get_service_keys().key, msg.signature);
+
+    msg.type = pulse::message_type::handshake;
+  }
+
+  pulse_relay_message_to_quorum(self, msg, quorum);
+}
+
+void send_pulse_validator_handshake_bit(void *self, service_nodes::quorum const &quorum, crypto::hash const &top_hash)
+{
+  send_pulse_validator_handshake_bit_or_bitset(self, false /*sending_bitset*/, quorum, top_hash, 0);
+}
+
+void send_pulse_validator_handshake_bitset(void *self, service_nodes::quorum const &quorum, crypto::hash const &top_hash, uint16_t validator_bitset)
+{
+  send_pulse_validator_handshake_bit_or_bitset(self, true /*sending_bitset*/, quorum, top_hash, validator_bitset);
+}
+
+// Invoked when daemon has received a participation handshake message via
+// QuorumNet from another validator, either forwarded or originating from that
+// node. The message is added to the Pulse message queue and validating the
+// contents of the message is left to the caller.
+void handle_pulse_participation_bit_or_bitset(Message &m, QnetState& qnet, bool bitset)
+{
+  if (m.data.size() != 1)
+      throw std::runtime_error(std::string("Rejecting pulse participation ") + (bitset ? std::string("bitset") : std::string("handshake")) + ": expected one data entry not " + std::to_string(m.data.size()));
+
+  int quorum_position         = -1;
+  uint16_t validator_bitset   = 0;
+  crypto::signature signature = {};
+
+  // NOTE: Extract Data
+  bt_dict_consumer data{m.data[0]};
+  if (bitset)
+  {
+    std::string_view INVALID_ARG_PREFIX = bitset ? "Invalid pulse validator bitset: missing required field '"sv
+                                                 : "Invalid pulse validator bit: missing required field '"sv;
+    if (data.skip_until(PULSE_TAG_VALIDATOR_BITSET))
+      validator_bitset = data.consume_integer<uint16_t>();
+    else
+      throw std::invalid_argument(std::string(INVALID_ARG_PREFIX) + std::string(PULSE_TAG_VALIDATOR_BITSET) + "'");
+
+    if (data.skip_until(PULSE_TAG_QUORUM_POSITION))
+      quorum_position = data.consume_integer<int>();
+    else
+      throw std::invalid_argument(std::string(INVALID_ARG_PREFIX) + std::string(PULSE_TAG_QUORUM_POSITION) + "'");
+
+    if (data.skip_until(PULSE_TAG_SIGNATURE)) {
+      auto sig_str = data.consume_string_view();
+      signature    = convert_string_view_bytes_to_signature(sig_str);
+    } else {
+      throw std::invalid_argument(std::string(INVALID_ARG_PREFIX) + std::string(PULSE_TAG_SIGNATURE) + "'");
+    }
+
+    assert(validator_bitset != -1);
+    assert(quorum_position != -1);
+    assert(signature);
```

Comment:
These should be throws rather than asserts since they are triggered by external input which could abort the program (admittedly only for a debug build).  The signature check shouldn't be necessary as the conversion function already throws for a null signature.